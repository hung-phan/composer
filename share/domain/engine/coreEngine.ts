import { Builder } from "builder-pattern";
import * as _ from "lodash";
import Router from "next/router";
import { DefaultRootState } from "react-redux";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";

import fetch from "../../library/fetch";
import {
  ClientInfo,
  DataContainer,
  Element,
  HttpMethod,
  HttpMethodRequestBody,
  InvokeExternalMethod,
  Method,
  NavigateMethod,
  RenderElementMethod,
  Response,
  UpdateElementMethod,
  UpdateInListElementMethod,
} from "../interfaces";
import { getSimplifiedElement } from "./coreEngineHelpers";
import getTaskQueue, { TaskQueue } from "./coreEngineQueue";
import { actions, selectors } from "./coreEngineStateStore";
import { decode } from "./serializers";

class CoreEngine {
  readonly taskQueue: TaskQueue;
  readonly dispatch: ThunkDispatch<DefaultRootState, unknown, AnyAction>;
  readonly clientInfo?: ClientInfo<any>;
}

async function evalMethod(
  method: Method,
  coreEngine: CoreEngine
): Promise<void> {
  if (method === null) {
    return;
  }

  if (method instanceof InvokeExternalMethod) {
    coreEngine.dispatch(method.data);
  } else if (method instanceof HttpMethod) {
    if (method.clientStateId !== undefined) {
      await coreEngine.dispatch((__, getState) =>
        makeHttpCall(
          method,
          coreEngine,
          selectors.getElementState(getState(), method.clientStateId)
        )
      );
    } else {
      await makeHttpCall(method, coreEngine);
    }
  } else if (method instanceof RenderElementMethod) {
    await registerElement(method.element, coreEngine);
  } else if (method instanceof UpdateElementMethod) {
    await coreEngine.dispatch(async (__, getState) => {
      const parentElement = selectors.getParentElement(getState(), method.id);

      await registerElement(method.element, coreEngine);

      coreEngine.dispatch(
        actions.replaceElement({
          parentId: parentElement.id,
          oldId: method.id,
          id: method.element.id,
        })
      );
    });
  } else if (method instanceof UpdateInListElementMethod) {
    await Promise.all(
      method.elements.map((element) => registerElement(element, coreEngine))
    );
    coreEngine.dispatch(
      actions.replaceElementInList({
        oldId: method.id,
        ids: method.elements.map((element) => element.id),
      })
    );
  } else if (method instanceof NavigateMethod) {
    if (process.env.ENVIRONMENT === "client") {
      await Router.push(method.url);
    } else {
      throw new Error(
        `Unsupported method for serverSide: ${method.interfaceName}`
      );
    }
  } else {
    throw new Error(`Unsupported method: ${method.interfaceName}`);
  }
}

async function makeHttpCall(
  method: HttpMethod<any>,
  coreEngine: CoreEngine,
  elementState?: DataContainer
): Promise<void> {
  if (method.requestType === undefined) {
    throw new Error("requestType cannot be undefined");
  }

  if (method.before) {
    await dispatchTask(
      coreEngine.taskQueue,
      coreEngine.dispatch,
      method.before,
      coreEngine.clientInfo
    );
  }

  try {
    const requestOptions: RequestInit = {};
    const requestBody: HttpMethodRequestBody<any, any> = {
      elementState,
      clientInfo: coreEngine.clientInfo,
      requestData: method.requestData,
    };
    let serverUrl = method.url;

    if (method.requestType === "GET") {
      serverUrl += `?${Object.entries(requestBody)
        .filter(([, value]) => value !== undefined)
        .map(
          ([key, value]) =>
            `${key}=${encodeURIComponent(JSON.stringify(value))}`
        )
        .join("&")}`;
    } else if (method.requestType === "POST") {
      requestOptions.body = JSON.stringify(requestBody);
    }

    const response = await fetch(serverUrl, {
      ...requestOptions,
      method: method.requestType,
    });

    if (response.ok) {
      const responseObj = decode<Response>(await response.text());

      if (method.onSuccess) {
        await dispatchTask(
          coreEngine.taskQueue,
          coreEngine.dispatch,
          method.onSuccess,
          coreEngine.clientInfo
        );
      }

      await dispatchTask(
        coreEngine.taskQueue,
        coreEngine.dispatch,
        responseObj.methods,
        coreEngine.clientInfo
      );
    } else if (method.onError) {
      await dispatchTask(
        coreEngine.taskQueue,
        coreEngine.dispatch,
        method.onError,
        coreEngine.clientInfo
      );
    }
  } catch (e) {
    console.error(e);

    if (method.onError) {
      await dispatchTask(
        coreEngine.taskQueue,
        coreEngine.dispatch,
        method.onError,
        coreEngine.clientInfo
      );
    }
  }

  if (method.after) {
    await dispatchTask(
      coreEngine.taskQueue,
      coreEngine.dispatch,
      method.after,
      coreEngine.clientInfo
    );
  }
}

async function registerElement(
  element: Element,
  coreEngine: CoreEngine
): Promise<void> {
  const childElements: Element[] = [];

  for (const [key, value] of Object.entries(element)) {
    if (value instanceof Element) {
      await coreEngine.taskQueue.run(registerElement, value, coreEngine);

      element[key] = getSimplifiedElement(value);

      childElements.push(element[key]);
    }

    if (
      _.isArray(value) &&
      _.every(value, (nestedElement) => nestedElement instanceof Element)
    ) {
      await Promise.all(
        value.map((nestedElement) =>
          coreEngine.taskQueue.run(registerElement, nestedElement, coreEngine)
        )
      );

      element[key] = value.map((nestedElement) =>
        getSimplifiedElement(nestedElement)
      );
      childElements.push(...element[key]);
    }
  }

  coreEngine.dispatch(
    actions.setElement({
      element,
    })
  );

  if (!_.isEmpty(childElements)) {
    coreEngine.dispatch(
      actions.registerParent({
        parentId: element.id,
        ids: childElements.map((nestedElement) => nestedElement.id),
      })
    );
  }
}

async function dispatchTask(
  taskQueue: TaskQueue,
  dispatch: ThunkDispatch<DefaultRootState, unknown, AnyAction>,
  methods?: Method[],
  clientInfo?: ClientInfo<any>
) {
  if (_.isEmpty(methods)) {
    return;
  }

  const coreEngine = Builder(CoreEngine)
    .taskQueue(taskQueue)
    .dispatch(dispatch)
    .clientInfo(clientInfo)
    .build();

  try {
    await Promise.all(methods.map((method) => evalMethod(method, coreEngine)));
  } catch (e) {
    console.error(e);
  }
}

export function engineDispatch(
  dispatch: ThunkDispatch<DefaultRootState, unknown, AnyAction>,
  methods?: Method[],
  clientInfo?: ClientInfo<any>
): Promise<void> {
  if (_.isEmpty(methods)) {
    return;
  }

  return dispatchTask(getTaskQueue(), dispatch, methods, clientInfo);
}
