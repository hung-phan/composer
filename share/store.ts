import produce from "immer";
import { HYDRATE, MakeStore, createWrapper } from "next-redux-wrapper";
import { DefaultRootState } from "react-redux";
import {
  AnyAction,
  Store,
  applyMiddleware,
  combineReducers,
  compose,
  createStore,
} from "redux";
import thunkMiddleware, { ThunkMiddleware } from "redux-thunk";

import {
  State as TemplateEngineState,
  selectors,
  mountPoint as templateEngineMountPoint,
  reducer as templateEngineReducer,
} from "./domain/engine/coreEngineStateStore";
import { convertToClass, encode } from "./domain/engine/serializers";
import { Element, Node } from "./domain/interfaces";

export const middlewares = [
  thunkMiddleware as ThunkMiddleware<DefaultRootState, AnyAction>,
];

export const enhancers = [];

if (
  process.env.ENVIRONMENT === "client" &&
  process.env.NODE_ENV === "development"
) {
  if ((window as any).__REDUX_DEVTOOLS_EXTENSION__) {
    enhancers.push((window as any).__REDUX_DEVTOOLS_EXTENSION__());
  }
}

export interface RootState {
  [templateEngineMountPoint]: TemplateEngineState;
}

const applicationReducer = combineReducers({
  [templateEngineMountPoint]: templateEngineReducer,
});

export const makeStore: MakeStore<Store<RootState>> = () =>
  createStore(
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    (state: RootState, action: AnyAction) =>
      produce(state, (draft) => {
        switch (action.type) {
          case HYDRATE:
            Object.assign(draft, action.payload);
            break;

          default:
            return applicationReducer(draft, action);
        }
      }),
    { [templateEngineMountPoint]: {} }, // preloadedState
    compose(applyMiddleware(...middlewares), ...enhancers)
  );

export const wrapper = createWrapper<Store<RootState>>(makeStore, {
  serializeState: encode,
  deserializeState: (encodedData: string) => {
    const state = JSON.parse(encodedData);

    const templateEngineState = selectors.getState(state);

    for (const [key, value] of Object.entries(templateEngineState)) {
      templateEngineState[key].element = convertToClass<Element>(value.element);
      templateEngineState[key] = convertToClass<Node>(value);
    }

    return state;
  },
});
