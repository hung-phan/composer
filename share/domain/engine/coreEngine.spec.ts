import { makeStore } from "../../store";
import {
  ButtonElement,
  ImageElement,
  LayoutElement,
  PlaceholderElement,
  RenderElementMethod,
  StateHolderElement,
  TextElement,
  UpdateElementMethod, UpdateInListElementMethod,
} from "../interfaces";
import { engineDispatch } from "./coreEngine";
import { ROOT_ID, actions } from "./index";

describe("coreEngine", () => {
  test("engineDispatch should update the store correctly", async () => {
    const store = makeStore({});

    await engineDispatch(store.dispatch, [
      RenderElementMethod.builder()
        .element(
          LayoutElement.builder()
            .id(ROOT_ID)
            .elements([
              TextElement.builder().id("text_element").build(),
              ImageElement.builder()
                .id("image_element")
                .src("https://google.com")
                .build(),
              ButtonElement.builder().id("button_element").build(),
              PlaceholderElement.builder().id("placeholder_element").build(),
              StateHolderElement.builder()
                .id("stateholder_element")
                .elementState({ data: ["string_data"] })
                .build(),
            ])
            .build()
        )
        .build(),
    ]);

    expect(store.getState()).toMatchSnapshot();

    await engineDispatch(store.dispatch, [
      RenderElementMethod.builder()
        .element(
          ImageElement.builder()
            .id("image_element")
            .src("https://amazon.com")
            .build()
        )
        .build(),
    ]);

    expect(store.getState()).toMatchSnapshot();

    await engineDispatch(store.dispatch, [
      RenderElementMethod.builder()
        .element(
          LayoutElement.builder()
            .id("placeholder_element")
            .elements([
              TextElement.builder()
                .id("new_text_element")
                .message("new text element message")
                .build(),
              PlaceholderElement.builder()
                .id("nested_placeholder_element_1")
                .build(),
              PlaceholderElement.builder()
                .id("nested_placeholder_element_2")
                .build(),
            ])
            .build()
        )
        .build(),
    ]);

    store.dispatch(
      actions.delElement({
        id: "placeholder_element",
        interfaceName: PlaceholderElement.getInterfaceName(),
      })
    );

    expect(store.getState()).toMatchSnapshot();

    await engineDispatch(store.dispatch, [
      UpdateElementMethod.builder()
        .id("nested_placeholder_element_1")
        .element(
          ImageElement.builder()
            .id("new_image_element")
            .build()
        )
        .build(),
    ]);

    store.dispatch(
      actions.delElement({
        id: "nested_placeholder_element_1",
        interfaceName: PlaceholderElement.getInterfaceName(),
      })
    );

    expect(store.getState()).toMatchSnapshot();

    await engineDispatch(store.dispatch, [
      UpdateInListElementMethod.builder()
        .id("nested_placeholder_element_2")
        .elements([
          ImageElement.builder()
            .id("new_image_element_1")
            .build(),
          ImageElement.builder()
            .id("new_image_element_2")
            .build()
        ])
        .build(),
    ]);

    store.dispatch(
      actions.delElement({
        id: "nested_placeholder_element_2",
        interfaceName: PlaceholderElement.getInterfaceName(),
      })
    );

    expect(store.getState()).toMatchSnapshot();
  });
});
