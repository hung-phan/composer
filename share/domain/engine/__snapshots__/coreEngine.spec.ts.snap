// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coreEngine should remove element with RenderElementMethod 1`] = `
Object {
  "templateEngine": Object {
    "@@ENGINE_ROOT_ID": Node {
      "childs": Object {
        "image_element": true,
        "placeholder_element": true,
      },
      "element": LayoutElement {
        "elements": Array [
          Element {
            "id": "image_element",
            "interfaceName": "ImageElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "placeholder_element",
            "interfaceName": "PlaceholderElement",
            Symbol(immer-draftable): true,
          },
        ],
        "id": "@@ENGINE_ROOT_ID",
        "interfaceName": "LayoutElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "button_element": Node {
      "element": ButtonElement {
        "id": "button_element",
        "interfaceName": "ButtonElement",
        "size": "lg",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "image_element": Node {
      "element": ImageElement {
        "id": "image_element",
        "interfaceName": "ImageElement",
        "src": "https://google.com",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "placeholder_element": Node {
      "element": PlaceholderElement {
        "id": "placeholder_element",
        "interfaceName": "PlaceholderElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "text_element": Node {
      "element": TextElement {
        "id": "text_element",
        "interfaceName": "TextElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
  },
}
`;

exports[`coreEngine should support list replacement with same id 1`] = `
Object {
  "templateEngine": Object {
    "@@ENGINE_ROOT_ID": Node {
      "childs": Object {
        "placeholder_element": true,
        "text_element_1": true,
        "text_element_2": true,
      },
      "element": LayoutElement {
        "elements": Array [
          Element {
            "id": "text_element_1",
            "interfaceName": "TextElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "text_element_2",
            "interfaceName": "TextElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "placeholder_element",
            "interfaceName": "PlaceholderElement",
            Symbol(immer-draftable): true,
          },
        ],
        "id": "@@ENGINE_ROOT_ID",
        "interfaceName": "LayoutElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "placeholder_element": Node {
      "element": PlaceholderElement {
        "id": "placeholder_element",
        "interfaceName": "PlaceholderElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "text_element_1": Node {
      "element": TextElement {
        "id": "text_element_1",
        "interfaceName": "TextElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "text_element_2": Node {
      "element": TextElement {
        "id": "text_element_2",
        "interfaceName": "TextElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
  },
}
`;

exports[`coreEngine should support nested element update with same id 1`] = `
Object {
  "templateEngine": Object {
    "@@ENGINE_ROOT_ID": Node {
      "childs": Object {
        "image_element_1": true,
        "nested_layout_element": true,
      },
      "element": LayoutElement {
        "elements": Array [
          Element {
            "id": "image_element_1",
            "interfaceName": "ImageElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "nested_layout_element",
            "interfaceName": "LayoutElement",
            Symbol(immer-draftable): true,
          },
        ],
        "id": "@@ENGINE_ROOT_ID",
        "interfaceName": "LayoutElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "image_element_1": Node {
      "element": ImageElement {
        "id": "image_element_1",
        "interfaceName": "ImageElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "image_element_2": Node {
      "element": ImageElement {
        "id": "image_element_2",
        "interfaceName": "ImageElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "nested_layout_element",
      Symbol(immer-draftable): true,
    },
    "nested_layout_element": Node {
      "childs": Object {
        "image_element_2": true,
        "placeholder_element": true,
      },
      "element": LayoutElement {
        "elements": Array [
          Element {
            "id": "image_element_2",
            "interfaceName": "ImageElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "placeholder_element",
            "interfaceName": "PlaceholderElement",
            Symbol(immer-draftable): true,
          },
        ],
        "id": "nested_layout_element",
        "interfaceName": "LayoutElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "placeholder_element": Node {
      "element": PlaceholderElement {
        "id": "placeholder_element",
        "interfaceName": "PlaceholderElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "nested_layout_element",
      Symbol(immer-draftable): true,
    },
  },
}
`;

exports[`coreEngine should support override element with RenderElementMethod 1`] = `
Object {
  "templateEngine": Object {
    "@@ENGINE_ROOT_ID": Node {
      "childs": Object {
        "image_element": true,
        "placeholder_element": true,
      },
      "element": LayoutElement {
        "elements": Array [
          Element {
            "id": "image_element",
            "interfaceName": "ImageElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "placeholder_element",
            "interfaceName": "LayoutElement",
            Symbol(immer-draftable): true,
          },
        ],
        "id": "@@ENGINE_ROOT_ID",
        "interfaceName": "LayoutElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "button_element": Node {
      "element": ButtonElement {
        "id": "button_element",
        "interfaceName": "ButtonElement",
        "size": "lg",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "placeholder_element",
      Symbol(immer-draftable): true,
    },
    "image_element": Node {
      "element": ImageElement {
        "id": "image_element",
        "interfaceName": "ImageElement",
        "src": "https://google.com",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "placeholder_element": Node {
      "childs": Object {
        "button_element": true,
        "text_element": true,
      },
      "element": LayoutElement {
        "elements": Array [
          Element {
            "id": "text_element",
            "interfaceName": "TextElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "button_element",
            "interfaceName": "ButtonElement",
            Symbol(immer-draftable): true,
          },
        ],
        "id": "placeholder_element",
        "interfaceName": "LayoutElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "text_element": Node {
      "element": TextElement {
        "id": "text_element",
        "interfaceName": "TextElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "placeholder_element",
      Symbol(immer-draftable): true,
    },
  },
}
`;

exports[`coreEngine should update the store correctly 1`] = `
Object {
  "templateEngine": Object {
    "@@ENGINE_ROOT_ID": Node {
      "childs": Object {
        "button_element": true,
        "image_element": true,
        "placeholder_element": true,
        "stateholder_element": true,
        "text_element": true,
      },
      "element": LayoutElement {
        "elements": Array [
          Element {
            "id": "text_element",
            "interfaceName": "TextElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "image_element",
            "interfaceName": "ImageElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "button_element",
            "interfaceName": "ButtonElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "placeholder_element",
            "interfaceName": "PlaceholderElement",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "stateholder_element",
            "interfaceName": "StateHolderElement",
            Symbol(immer-draftable): true,
          },
        ],
        "id": "@@ENGINE_ROOT_ID",
        "interfaceName": "LayoutElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "button_element": Node {
      "element": ButtonElement {
        "id": "button_element",
        "interfaceName": "ButtonElement",
        "size": "lg",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "image_element": Node {
      "element": ImageElement {
        "id": "image_element",
        "interfaceName": "ImageElement",
        "src": "https://google.com",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "placeholder_element": Node {
      "element": PlaceholderElement {
        "id": "placeholder_element",
        "interfaceName": "PlaceholderElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "stateholder_element": Node {
      "element": StateHolderElement {
        "elementState": Object {
          "data": Array [
            "string_data",
          ],
        },
        "id": "stateholder_element",
        "interfaceName": "StateHolderElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "text_element": Node {
      "element": TextElement {
        "id": "text_element",
        "interfaceName": "TextElement",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
  },
}
`;
